File: lib/core/transformers/standard/transformer.dart
/// GPT-style decoder-only Transformer.
import 'package:dart_cuda/core/layers/layer_norm.dart';
import 'package:dart_cuda/core/layers/nn.dart';
import 'package:dart_cuda/core/tensor/gpu_tensor.dart';
import 'package:dart_cuda/core/transformers/standard/transformer_block.dart';

class Transformer extends Module {
  final int vocabSize;
  final int embedSize;
  final int blockSize;
  final int numLayers;
  final int numHeads;

  final Tensor wte; // [vocabSize, embedSize]
  final Tensor wpe; // [blockSize, embedSize]
  final List<TransformerBlock> blocks;
  final LayerNorm finalLayerNorm;
  final Layer lmHead;

  Transformer({
    this.vocabSize = 50,
    this.embedSize = 32,
    this.blockSize = 8,
    this.numLayers = 4,
    this.numHeads = 4,
  })  : assert(embedSize % numHeads == 0,
            'embedSize must be divisible by numHeads'),
        wte = Tensor.random([vocabSize, embedSize], scale: 0.02),
        wpe = Tensor.random([blockSize, embedSize], scale: 0.02),
        blocks = List.generate(
          numLayers,
          (_) => TransformerBlock(embedSize, numHeads, masked: true),
        ),
        finalLayerNorm = LayerNorm(embedSize),
        lmHead = Layer(embedSize, vocabSize, useGelu: false);

  Tensor forward(List<int> idx, List<Tensor> tracker) {
    final T = idx.length;
    if (T > blockSize) {
      throw ArgumentError(
          'Sequence length $T exceeds block size $blockSize');
    }
    var x = Tensor.embeddings(idx, wte, wpe);
    tracker.add(x);
    for (final b in blocks) {
      x = b.forward(x, tracker);
    }
    x = finalLayerNorm.forward(x, tracker);
    return lmHead.forward(x, tracker);
  }

  @override
  List<Tensor> parameters() => [
        wte,
        wpe,
        ...blocks.expand((b) => b.parameters()),
        ...finalLayerNorm.parameters(),
        ...lmHead.parameters(),
      ];
}

File: python/dart_cuda/core/transformers/standard/transformer.py
"""GPT-style decoder-only Transformer — port of
`lib/transformer/transformer.dart` (dart-torch).
"""

from __future__ import annotations

from typing import List, Sequence

from ...layers.layer_norm import LayerNorm
from ...layers.nn import Layer, Module
from ...tensor.gpu_tensor import Tensor
from .transformer_block import TransformerBlock


class Transformer(Module):
    def __init__(
        self,
        *,
        vocab_size: int = 50,
        embed_size: int = 32,
        block_size: int = 8,
        num_layers: int = 4,
        num_heads: int = 4,
    ):
        assert embed_size % num_heads == 0
        self.vocab_size = vocab_size
        self.embed_size = embed_size
        self.block_size = block_size
        self.num_layers = num_layers
        self.num_heads = num_heads

        self.wte = Tensor.random([max(vocab_size, 1), embed_size], scale=0.02)
        self.wpe = Tensor.random([block_size, embed_size], scale=0.02)
        self.blocks = [
            TransformerBlock(embed_size, num_heads, masked=True)
            for _ in range(num_layers)
        ]
        self.final_layer_norm = LayerNorm(embed_size)
        self.lm_head = Layer(embed_size, vocab_size, useGelu=False)

    # dart-torch aliases
    @property
    def vocabSize(self) -> int: return self.vocab_size
    @property
    def embedSize(self) -> int: return self.embed_size
    @property
    def blockSize(self) -> int: return self.block_size
    @property
    def numLayers(self) -> int: return self.num_layers
    @property
    def numHeads(self) -> int: return self.num_heads
    @property
    def finalLayerNorm(self) -> LayerNorm: return self.final_layer_norm
    @property
    def lmHead(self) -> Layer: return self.lm_head

    def forward(self, idx: Sequence[int], tracker: List[Tensor]) -> Tensor:
        T = len(idx)
        if T > self.block_size:
            raise ValueError(
                f"Sequence length {T} exceeds block size {self.block_size}"
            )
        x = Tensor.embeddings(idx, self.wte, self.wpe)
        tracker.append(x)
        for b in self.blocks:
            x = b.forward(x, tracker)
        x = self.final_layer_norm.forward(x, tracker)
        logits = self.lm_head.forward(x, tracker)
        return logits

    def parameters(self) -> List[Tensor]:
        ps: List[Tensor] = [self.wte, self.wpe]
        for b in self.blocks:
            ps.extend(b.parameters())
        ps.extend(self.final_layer_norm.parameters())
        ps.extend(self.lm_head.parameters())
        return ps
